### PR TITLE
Simplified Feign

### LINKED ISSUE #202: Simplify Feign integration
We've hidden all the impl details of https://github.com/spring-cloud/spring-cloud-sleuth/commit/19249cdc23a655125a58d022525276e1f74aae09 to make it possible to instrument Feign in a nicer way in the future.

CC @adriancole @spencergibb

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
with this change we no longer treat retries as a continuation of a previous span. That way the Feign code simplifies a lot. RequestInterceptor starts a span and the TraceFeignClient will always close it no matter what's happening.

fixes #202

cc @adriancole

### CODE REVIEW

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignClient.java [hunk 1]
```diff
@@ -36,42 +39,34 @@
...
 	}
 
 	@Override
 	public Response execute(Request request, Request.Options options) throws IOException {
-		Response response;
 		try {
 			addRequestTags(request);
-			response = this.delegate.execute(request, options);
+			Response response = this.delegate.execute(request, options);
+			logCr();
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignClient.java]: I'm logging CR only after successful response

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignClient.java [hunk 2]
```diff
@@ -36,42 +39,34 @@
...
+			return response;
+		} finally {
+			closeSpan();
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignClient.java]: If there was an success / exception the span will be closed. Only upon successful response will the span get CR logged

Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignClient.java]: on IOException|RuntimeException you could add a binary annotation of "error" -> message, as that will turn the thing red in zipkin's UI

Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignClient.java]: Nice idea

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignClientAutoConfiguration.java [hunk 3]
```diff
@@ -104,26 +104,26 @@ TraceFeignObjectWrapper traceFeignObjectWrapper(BeanFactory beanFactory) {
...
+		return new ResponseEntityDecoder(new SpringDecoder(this.messageConverters)) {
+				@Override
+				public Object decode(Response response, Type type)
+						throws IOException, FeignException {
+					Tracer tracer = beanFactory.getBean(Tracer.class);
+					if (tracer == null) {
+						throw new IllegalStateException("Trace hasn't been yet initialized");
 					}
-				});
+					FeignResponseHeadersHolder feignResponseHeadersHolder = new FeignResponseHeadersHolder(
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignClientAutoConfiguration.java]: why are we injecting headers into the response?

**Comment by adriancole** [comment]
looks good!

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignClientAutoConfiguration.java (hunk 3 above)
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignClientAutoConfiguration.java]: Actually that was an approach that we agreed on not doing some time ago... Good catch - I removed it

Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignClientAutoConfiguration.java]: Ah! It's this old approach where we injected trace stuff into the response. I think it should be deleted. Good catch!

**Pull request merged**